lib/chef/knife/ec_key_import.rb

```diff
@@ -39,17 +40,130 @@ class EcKeyImport < Chef::Knife
         :boolean => true,
         :description => "Upload user ids."
 
+      option :users_only,
+        :long => "--users-only",
+        :default => false,
+        :description => "Only update users (skip client key data)"
+
+     option :clients_only,
+        :log => "--clients-only",
+        :default => false,
+        :description => "Only update client key data. Implies --skip-users-table"
+
       def run
         if config[:sql_user].nil? || config[:sql_password].nil?
           load_config_from_file!
         end
 
-        path = @name_args[0] || "key_dump.json"
-        import(path)
+        @org_id_cache = Chef::OrgIdCache.new(db)
+
+        # user_data_path defaults to key_dump.json to support
+        # older knife-ec-backup exports
+        user_data_path = @name_args[0] || "key_dump.json"
+        key_data_path = @name_args[1] || "key_table_dump.json"
+        import_user_data(user_data_path) unless (config[:skip_users_table] || config[:clients_only])
+        import_key_data(key_data_path) unless config[:skip_keys_table]
       end
 
+      def import_key_data(path)
+        key_data = JSON.parse(File.read(path))
+        key_data.each do |d|
+          case d['type']
+          when 'client'
+            next if config[:users_only]
+            insert_key_data_for_client(d)
+          when 'user'
+            next if config[:clients_only]
+            insert_key_data_for_user(d)
+          else
+            ui.warn "Unkown actor type #{d['type']} for #{d['name']}"
+            next
+          end
+        end
+      end
+
+      # If a given key_name already exists for the client, update it,
+      # otherwise insert a new key into the key table.
+      #
+      # Unlike users, we never want to keep the internal ID from the
+      # backup.
+      #
+      # The org_id is likely different than that stored in the backup.
+      # We query the new org_id based on org_name, caching it to avoid
+      # multiple lookups in a large org.
+      def insert_key_data_for_client(d)
+        ui.msg "Updating key data for client[#{d['name']}]"
+
+        org_id = @org_id_cache.fetch(d['org_name'])
+        if org_id.nil?
+          ui.warn "Could not find organization for client[#{d['name']}], skipping."
+          ui.warn "Organizations much be restored before key data can be imported."
+          return
+        end
+
+        existing_client = db[:clients].where(:org_id => org_id, :name => d['name']).first
+        if existing_client.nil?
+          ui.warn "Did not find existing client record for #{d['name']}, skipping."
+          ui.warn "Client records must be restored before key data can be imported."
+          return
+        end
+
+        new_id = existing_client = new_id
```

Comment:
This should be `new_id = existing_client[:id]`.  I am now very confused how this is working.